File: go/stack/leetcode/next_greater_element_i.go
package main

//下一个更大的元素I

//暴力法
func nextGreaterElement1(nums1 []int, nums2 []int) []int {
	var res []int

	m := make(map[int]int)
	for i, v := range nums2 {
		m[v] = i
	}

	for _, v := range nums1 {
		exist := false
		for j := m[v] + 1; j < len(nums2); j++ {
			if nums2[j] > v {
				exist = true
				res = append(res, nums2[j])
				break
			}
		}
		if !exist {
			res = append(res, -1)
		}
	}
	return res
}

//使用一个栈+哈希表
func nextGreaterElement2(nums1 []int, nums2 []int) []int {
	var stack []int
	var res []int
	m := make(map[int]int)

	for _, v := range nums2 {
		for len(stack) > 0 && stack[len(stack)-1] < v {
			m[stack[len(stack)-1]] = v
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, v)
	}

	for _, v := range nums1 {
		if _, exist := m[v]; exist {
			res = append(res, m[v])
		} else {
			res = append(res, -1)
		}
	}
	return res
}

File: python/stack/leetcode/next_greater_element_i.py

# 下一个更大元素I
from typing import List


class Solution:

    # 暴力法
    def nextGreaterElement1(self, nums1: List[int], nums2: List[int]) -> List[int]:
        m, res = {}, []
        for i in range(len(nums2)):
            m[nums2[i]] = i

        for i in range(len(nums1)):
            exist = False
            for j in range(m[nums1[i]] + 1, len(nums2)):
                if nums2[j] > nums1[i]:
                    exist = True
                    res.append(nums2[j])
                    break
            if not exist:
                res.append(-1)
        return res

    # 使用栈+哈希表
    def nextGreaterElement2(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack, m = [], {}
        for n in nums2:
            while len(stack) > 0 and stack[-1] < n:
                m[stack[-1]] = n
                stack = stack[:-1]
            stack.append(n)

        return [m.get(n, -1) for n in nums1]
